brian2/devices/cpp_standalone/device.py

```diff
@@ -766,18 +767,23 @@ def copy_source_files(self, writer, directory):
                                   'random', 'randomkit', 'randomkit.h'),
                      os.path.join(directory, 'brianlib', 'randomkit', 'randomkit.h'))
 
+    def _insert_func_namespace(self, func, code_object, namespace):
+        impl = func.implementations[CPPStandaloneCodeObject]
+        func_namespace = impl.get_namespace(code_object.owner)
+        if func_namespace is not None:
+            namespace.update(func_namespace)
+        if impl.dependencies is not None:
+            for dep in impl.dependencies.itervalues():
+                self._insert_func_namespace(dep, code_object, namespace)
+
     def write_static_arrays(self, directory):
-        # # Find np arrays in the namespaces and convert them into static
-        # # arrays. Hopefully they are correctly used in the code: For example,
-        # # this works for the namespaces for functions with C++ (e.g. TimedArray
-        # # treats it as a C array) but does not work in places that are
-        # # implicitly vectorized (state updaters, resets, etc.). But arrays
-        # # shouldn't be used there anyway.
+        # Write Function namespaces as static arrays
         for code_object in self.code_objects.itervalues():
-            for name, value in code_object.variables.iteritems():
-                if isinstance(value, np.ndarray):
-                    self.static_arrays[name] = value
-                    
```

Comment:
More or less: it is part of a fundamental change how function namespaces (most important use case is `TimedArray`) where added. Previously, we added them to the original `variables` dictionary at some point. This screwed up caching, because the `variables` were used as part of the cache key earlier and therefore should not change. For this specific bit of code we therefore no longer search for those arrays as part of the `variables` dictionary, but explicitly add them from the `Function`.